### PR TITLE
[reflectcpp] Add open-source library reflectcpp to vcpkg

### PR DESCRIPTION
**Pull request by Urfoex**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
  - https://repology.org/project/reflect-cpp/versions 
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by Urfoex** [comment]
@microsoft-github-policy-service agree company="Code17 GmbH"

**Review by BillyONeal — DISMISSED** [review]
This port vendors dependencies which causes it to conflict with libraries already available in vcpkg. Please declare these as dependencies and use the versions provided by the corresponding ports.

```console
PS D:\vcpkg> .\vcpkg.exe install reflectcpp
Computing installation plan...
The following packages will be built and installed:
    reflectcpp:x64-windows@0.10.0
Detecting compiler hash for triplet x64-windows...
Compiler found: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe
Restored 0 package(s) from C:\Users\bion\AppData\Local\vcpkg\archives in 84.6 us. Use --debug to see more details.
Installing 1/1 reflectcpp:x64-windows@0.10.0...
Building reflectcpp:x64-windows@0.10.0...
-- Using cached getml-reflect-cpp-v0.10.0.tar.gz.
-- Cleaning sources at D:/vcpkg/buildtrees/reflectcpp/src/v0.10.0-7d3f399a13.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source D:/vcpkg-downloads/getml-reflect-cpp-v0.10.0.tar.gz
-- Using source at D:/vcpkg/buildtrees/reflectcpp/src/v0.10.0-7d3f399a13.clean
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Installing: D:/vcpkg/packages/reflectcpp_x64-windows/share/reflectcpp/copyright
-- Performing post-build validation
error: The following files are already installed in D:/vcpkg/installed/x64-windows and are in conflict with reflectcpp:x64-windows
Installed by ctre:x64-windows  
include/ctre.hpp

Installed by yyjson:x64-windows  
include/yyjson.h

Stored binaries in 1 destinations in 131 ms.
Elapsed time to handle reflectcpp:x64-windows: 6.7 
```

File: ports/reflectcpp/portfile.cmake [hunk 1]
```diff
@@ -6,7 +6,11 @@ vcpkg_from_github(
     HEAD_REF main
 )
 
-string(COMPARE EQUAL ${VCPKG_LIBRARY_LINKAGE} "dynamic" REFLECTCPP_BUILD_SHARED)
+if("${VCPKG_LIBRARY_LINKAGE}" STREQUAL "static" OR "${VCPKG_TARGET_IS_WINDOWS}")
```
Comment (jimwang118) [inline ports/reflectcpp/portfile.cmake:9]: I guess you want to set REFLECTCPP_BUILD_SHARED=OFF when building a static library under Windows, and set REFLECTCPP_BUILD_SHARED=ON if you compile it as a dynamic library. If this is the case, you should modify the judgment condition to if("${VCPKG_LIBRARY_LINKAGE}" STREQUAL "static" AND "${VCPKG_TARGET_IS_WINDOWS}").

Comment (dg0yt) [reply ports/reflectcpp/portfile.cmake:9]: I guess the package doesn't handle dllexport, so the author really wants to always turn off REFLECTCPP_BUILD_SHARED on Windows...

File: ports/reflectcpp/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    REF "v${VERSION}"
+    SHA512 f3d1d3b8f69fc0f93268586ae2002ef5b2646c1459333ab976bad05cdc02f57395917ff859bd67cc00f71495dd8d705027bea6d1011e5cd48d4bd78f2f164f45
+    HEAD_REF main
+)
+
+if("${VCPKG_LIBRARY_LINKAGE}" STREQUAL "static" OR "${VCPKG_TARGET_IS_WINDOWS}")
+    set(REFLECTCPP_BUILD_SHARED OFF)
+else()
+    set(REFLECTCPP_BUILD_SHARED ON)
+endif()
```
Comment (dg0yt) [inline ports/reflectcpp/portfile.cmake:13]: But in that case:
```suggestion
if(VCPKG_TARGET_IS_WINDOWS)
    vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
endif()
string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "dynamic" REFLECTCPP_BUILD_SHARED)
```

Comment (Urfoex) [reply ports/reflectcpp/portfile.cmake]: Changed.

**Comment by Urfoex** [comment]
> > This port vendors dependencies which causes it to conflict with libraries already available in vcpkg. Please declare these as dependencies and use the versions provided by the corresponding ports.
> > ```
> > PS D:\vcpkg> .\vcpkg.exe install reflectcpp
> > Computing installation plan...
> > The following packages will be built and installed:
> >     reflectcpp:x64-windows@0.10.0
> > Detecting compiler hash for triplet x64-windows...
> > Compiler found: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe
> > Restored 0 package(s) from C:\Users\bion\AppData\Local\vcpkg\archives in 84.6 us. Use --debug to see more details.
> > Installing 1/1 reflectcpp:x64-windows@0.10.0...
> > Building reflectcpp:x64-windows@0.10.0...
> > -- Using cached getml-reflect-cpp-v0.10.0.tar.gz.
> > -- Cleaning sources at D:/vcpkg/buildtrees/reflectcpp/src/v0.10.0-7d3f399a13.clean. Use --editable to skip cleaning for the packages you specify.
> > -- Extracting source D:/vcpkg-downloads/getml-reflect-cpp-v0.10.0.tar.gz
> > -- Using source at D:/vcpkg/buildtrees/reflectcpp/src/v0.10.0-7d3f399a13.clean
> > -- Configuring x64-windows
> > -- Building x64-windows-dbg
> > -- Building x64-windows-rel
> > -- Installing: D:/vcpkg/packages/reflectcpp_x64-windows/share/reflectcpp/copyright
> > -- Performing post-build validation
> > error: The following files are already installed in D:/vcpkg/installed/x64-windows and are in conflict with reflectcpp:x64-windows
> > Installed by ctre:x64-windows  
> > include/ctre.hpp
> > 
> > Installed by yyjson:x64-windows  
> > include/yyjson.h
> > 
> > Stored binaries in 1 destinations in 131 ms.
> > Elapsed time to handle reflectcpp:x64-windows: 6.7 
> > ```
> 
> We'll do upstream changes via: [getml/reflect-cpp#79](https://github.com/getml/reflect-cpp/issues/79) And update this PR afterwards.

Changed.

Should be ready to review again.

File: ports/reflectcpp/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+vcpkg_cmake_config_fixup(
+    CONFIG_PATH "lib/cmake/${PORT}"
+)
+
+file(REMOVE_RECURSE
+    "${CURRENT_PACKAGES_DIR}/debug"
+    "${CURRENT_PACKAGES_DIR}/lib"
+)
+
+file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
```
Comment (jimwang118) [inline ports/reflectcpp/portfile.cmake:36]:
```suggestion
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```

File: ports/reflectcpp/reflectcpp.0.11.1.patch [hunk 4]
```diff
@@ -0,0 +1,330 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
```
Comment (jimwang118) [inline ports/reflectcpp/reflectcpp.0.11.1.patch:1]: Does the fix of `reflectcpp.0.11.1.patch` patch need to be synchronized to upstream?

**Review by jimwang118 — CHANGES_REQUESTED** [review]

File: ports/reflectcpp/portfile.cmake (hunk 3 above)
Comment (Urfoex) [reply ports/reflectcpp/portfile.cmake]: Adjusted.

File: ports/reflectcpp/reflectcpp.0.11.1.patch (hunk 4 above)
Comment (Urfoex) [reply ports/reflectcpp/reflectcpp.0.11.1.patch]: Changes upstreamed.

**Review by jimwang118 — DISMISSED** [review]

**Comment by Urfoex** [comment]
> > This port vendors dependencies which causes it to conflict with libraries already available in vcpkg. Please declare these as dependencies and use the versions provided by the corresponding ports.
> > ```
> > PS D:\vcpkg> .\vcpkg.exe install reflectcpp
> > Computing installation plan...
> > The following packages will be built and installed:
> >     reflectcpp:x64-windows@0.10.0
> > Detecting compiler hash for triplet x64-windows...
> > Compiler found: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe
> > Restored 0 package(s) from C:\Users\bion\AppData\Local\vcpkg\archives in 84.6 us. Use --debug to see more details.
> > Installing 1/1 reflectcpp:x64-windows@0.10.0...
> > Building reflectcpp:x64-windows@0.10.0...
> > -- Using cached getml-reflect-cpp-v0.10.0.tar.gz.
> > -- Cleaning sources at D:/vcpkg/buildtrees/reflectcpp/src/v0.10.0-7d3f399a13.clean. Use --editable to skip cleaning for the packages you specify.
> > -- Extracting source D:/vcpkg-downloads/getml-reflect-cpp-v0.10.0.tar.gz
> > -- Using source at D:/vcpkg/buildtrees/reflectcpp/src/v0.10.0-7d3f399a13.clean
> > -- Configuring x64-windows
> > -- Building x64-windows-dbg
> > -- Building x64-windows-rel
> > -- Installing: D:/vcpkg/packages/reflectcpp_x64-windows/share/reflectcpp/copyright
> > -- Performing post-build validation
> > error: The following files are already installed in D:/vcpkg/installed/x64-windows and are in conflict with reflectcpp:x64-windows
> > Installed by ctre:x64-windows  
> > include/ctre.hpp
> > 
> > Installed by yyjson:x64-windows  
> > include/yyjson.h
> > 
> > Stored binaries in 1 destinations in 131 ms.
> > Elapsed time to handle reflectcpp:x64-windows: 6.7 
> > ```
> 
> We'll do upstream changes via: [getml/reflect-cpp#79](https://github.com/getml/reflect-cpp/issues/79) And update this PR afterwards.

Changes are upstreamed and integrated.

**Comment by jimwang118** [comment]
@BillyONeal Do you have any other modification requests for this PR?

**Comment by Urfoex** [comment]
> @BillyONeal Do you have any other modification requests for this PR?

Any news?

**Comment by BillyONeal** [comment]
Sorry I lost track of this.

**Comment by BillyONeal** [comment]
Closed and reopened to trigger a remerge/retest given that it's been a while

**Comment by BillyONeal** [comment]
Thanks!

**Pull request merged**